File: EFTAIOS/outdated-code/HexagonShape.java
//package view;
//
//import javafx.scene.paint.Color;
//import javafx.scene.shape.Polygon;
//import javafx.scene.shape.StrokeType;
//import model.map.MapNode;
//import model.map.NodeType;
//import model.GameClientController;
//
//
///* Mathematics of Hexagons (for personal reference). When creating a hexagon centered at x, y with a distance of
// * n between each of its corners:
// * recall that a 30-60-90 triangle with a hypotenuse of length n, the length of the side opposite 30 degrees
// * is n/2. The length of the side opposite to 60 degrees is sqrt(3)*n/2
// * */
//
///**
// * @author Daniel Cimento & Tristan Shoemaker
// * Huge thanks to Tristan for helping me debug my geometry calculations!
// * */
//public class HexagonShape extends Polygon{
//	private static final Color DARKESTGREY = Color.rgb(100, 100, 100);
//
//	String coordinate;
//
//	public HexagonShape(double xCoordinate, double yCoordinate, double size, String coordinateName, GameClientController client){
//		super(calculateCoordinates(xCoordinate, yCoordinate, size));
//
//		setFill(Color.WHITE);
//		setDefaultStroke();
//
//		coordinate = coordinateName;
//		setOnMouseClicked(e -> client.mapNodeSelected(MapNode.get(coordinate)));
//	}
//
//	public void setHexType(NodeType t){
//		switch (t) {
//		case ALIEN:
//			setFill(Color.RED);
//			break;
//		case HUMAN:
//			setFill(Color.PINK);
//			break;
//		case DANGEROUS:
//			setFill(Color.SILVER);
//			break;
//		case SECURE:
//			setFill(Color.WHITE);
//			break;
//		case ESCAPE_HATCH:
//			setFill(Color.GREEN);
//			break;
//		default:
//			setFill(Color.WHITE);
//			break;
//		}
//	}
//
//	public void setDefaultStroke(){
//		setStroke(DARKESTGREY);
//		setStrokeWidth(2.0);
//		setStrokeType(StrokeType.CENTERED);
//	}
//
//	public void setSelectedStroke(){
//		setStroke(Color.ORANGE);
//		setStrokeWidth(3.0);
//		setStrokeType(StrokeType.INSIDE);
//	}
//
//	public boolean isSelectedStroke(){
//		return getStroke().equals(Color.ORANGE) && getStrokeType().equals(StrokeType.INSIDE) && getStrokeWidth() == 3.0;
//	}
//
//	/**
//	 * With a hypotenuse of length n, this function returns the length of the longer side in a 30-60-90 triangle
//	 * */
//	private static double calculateLongSide(double n){
//		return ((Math.sqrt(3)*(n))/2);
//	}
//
//	private static double calculateShortSide(double n){
//		return n/2;
//	}
//
//	private static double[] calculateCoordinates(double xCoordinate, double yCoordinate, double n){
//		//hexagon has 6 coordinates, therefore 12 points are necessary
//		double[] coordinates = new double[12];
//		//first point is n to the right of the center
//		coordinates[0] = xCoordinate + n;
//		coordinates[1] = yCoordinate;
//		//second point is n/2 to the right and sqrt(3)n/2 up
//		coordinates[2] = xCoordinate + calculateShortSide(n);
//		coordinates[3] = yCoordinate + calculateLongSide(n);
//		//third point is " left and " up
//		coordinates[4] = xCoordinate - calculateShortSide(n);
//		coordinates[5] = yCoordinate + calculateLongSide(n);
//		//fourth point is n to the left
//		coordinates[6] = xCoordinate - n;
//		coordinates[7] = yCoordinate;
//
//		coordinates[8] = xCoordinate - calculateShortSide(n);
//		coordinates[9] = yCoordinate - calculateLongSide(n);
//
//		coordinates[10] = xCoordinate + calculateShortSide(n);
//		coordinates[11] = yCoordinate - calculateLongSide(n);
//
//		return coordinates;
//	}
//}

File: src/view/HexagonShape.scala
package view

import javafx.scene.paint.Color
import javafx.scene.shape.Polygon
import javafx.scene.shape.StrokeType

import model.actions.MoveAction
import model.engine.ActionListener
import model.map._

/* Mathematics of Hexagons (for personal reference). When creating a hexagon centered at x, y with a distance of
 * n between each of its corners:
 * recall that a 30-60-90 triangle with a hypotenuse of length n, the length of the side opposite 30 degrees
 * is n/2. The length of the side opposite to 60 degrees is sqrt(3)*n/2
 * */
/**
  * @author Daniel Cimento & Tristan Shoemaker
  *         Huge thanks to Tristan for helping me debug my geometry calculations!
  **/
object HexagonShape {
  private val DARKEST_GREY = Color.rgb(100, 100, 100)

  /**
    * With a hypotenuse of length n, this function returns the length of the longer side in a 30-60-90 triangle
    **/
  private def calculateLongSide(n: Double) = (Math.sqrt(3) * n) / 2

  private def calculateShortSide(n: Double) = n / 2

  private def calculateCoordinates(xCoordinate: Double, yCoordinate: Double, n: Double): List[Double] = {
    //hexagon has 6 coordinates, therefore 12 points are necessary
    List(
      // Point 1 is n to the right of the center
      xCoordinate + n,
      yCoordinate,
      // Point 2 is n/2 to the right and sqrt(3)n/2 up
      xCoordinate + calculateShortSide(n),
      yCoordinate + calculateLongSide(n),
      // Point 3 is " left and " up
      xCoordinate - calculateShortSide(n),
      yCoordinate + calculateLongSide(n),
      // Point 4 : The next 3 points are symmetric to the first 3
      xCoordinate - n,
      yCoordinate,
      // Point 5
      xCoordinate - calculateShortSide(n),
      yCoordinate - calculateLongSide(n),
      // Point 6
      xCoordinate + calculateShortSide(n),
      yCoordinate - calculateLongSide(n)
    )
  }
}

/**
  * Builds a hexagon for use with JavaFX
  * @param xCoordinate The center horizontal coordinate of the hexagon in the cartesian plane
  * @param yCoordinate The center vertical coordinate of the hexagon in the cartesian plane
  * @param size The "radius" of the hexagon, measured as the circle which circumscribes the hexagon
  * @param mapNode The model.map's node which this hexagon is meant to visually represent
  * @param actionListener The controller layer to allow callbacks on click
  */
class HexagonShape(
  val xCoordinate: Double,
  val yCoordinate: Double,
  val size: Double,
  val mapNode: MapNode,
  val actionListener: ActionListener
) extends Polygon(HexagonShape.calculateCoordinates(xCoordinate, yCoordinate, size): _*) {
  // Initialization Code
  {
    setFill(getHexColor(mapNode.nodeType))
    setDefaultStroke()
    setOnMouseClicked(e => actionListener.receiveAction(MoveAction(mapNode.name)))
  }

  private def getHexColor(t: NodeType): Color = {
    t match {
      case Alien => Color.RED
      case Human => Color.PINK
      case Dangerous => Color.SILVER
      case Secure | Blocked => Color.WHITE
      case EscapeHatch => Color.GREEN
    }
  }

  def setDefaultStroke() {
    setStroke(HexagonShape.DARKEST_GREY)
    setStrokeWidth(2.0)
    setStrokeType(StrokeType.CENTERED)
  }

  def setSelectedStroke() {
    setStroke(Color.ORANGE)
    setStrokeWidth(3.0)
    setStrokeType(StrokeType.INSIDE)
  }

  def isSelectedStroke: Boolean = getStroke == Color.ORANGE && getStrokeType == StrokeType.INSIDE && getStrokeWidth == 3.0
}
